### PR TITLE
Rename NGenHashTable to DacEnumerableHashTable to describe its unique feature and remove dead code

### LINKED ISSUE #58048: Rename/consolidate NgenHashTable in CoreCLR
In CoreCLR, we have a data structure called `NgenHashTable` that's used as the base of a number of our hash tables/caches in the runtime. Now that NGEN is dead, we should rename this type and remove any NGen-specific functionality that remains (such as hot/cold splitting and fixups)

cc: @AaronRobinsonMSFT @elinor-fung @davidwrighton

### PR DESCRIPTION
**Pull request by jkoritzinsky**
Remove the dead code from NGenHashTable now that NGEN is dead. Clean up the code around enumeration now that we have one set of buckets.

Use the C++ detection idiom (the C++11-compliant version) to allow implementing types to opt-in to special elementwise handling for DAC memory enumeration instead of requiring all subclasses to implement special handling , even when no special handling is required.

Additions to `type_traits` are pulled from the Microsoft STL and adapted for our code-style and the fact that we only use C++11.

Delete `GuidToMethodTableHashTable` as it is unused.

Fixes #58048

### CODE REVIEW

File: src/coreclr/inc/clr_std/type_traits [hunk 1]
```diff
@@ -571,6 +571,47 @@ namespace std
...
+        using _Rvalue = _Ty;
+    };
+
+    template <class _Ty>
+    struct _Add_reference<_Ty, detail::void_t<_Ty&>> { // (referenceable type)
+        using _Lvalue = _Ty&;
+        using _Rvalue = _Ty&&;
+    };
+
+    template <class _Ty>
```
[the remark this answers is not in the document]
Comment (jkoritzinsky) [reply src/coreclr/inc/clr_std/type_traits:600]: I added some of the add_lvalue_reference-related ones to make the snippet I pulled from the STL a complete snippet (so easier to update from changes in the STL, etc).

I can remove it if we would prefer.

File: src/coreclr/vm/dacenumerablehash.h [hunk 2]
```diff
@@ -0,0 +1,218 @@
...
+        return m_pBuckets;
+    }
+
+    // Loader heap provided at construction time. May be NULL (in which case m_pModule must *not* be NULL).
+    LoaderHeap             *m_pHeap;
+
+    // Fields related to the runtime (volatile or warm) part of the hash.
+    DPTR(PTR_VolatileEntry)                  m_pBuckets;  // Pointer to a simple bucket list (array of VolatileEntry pointers)
+    DWORD                                    m_cBuckets;  // Count of buckets in the above array (always non-zero)
+    DWORD                                    m_cEntries;  // Count of elements in the warm section of the hash
```
Comment (jkotas) [inline src/coreclr/vm/dacenumerablehash.h:215]:
```suggestion
    DWORD                                    m_cEntries;  // Count of elements
```

File: src/coreclr/vm/dacenumerablehash.h [hunk 3]
```diff
@@ -0,0 +1,218 @@
...
+    {
+        SUPPORTS_DAC;
+
+        return m_pBuckets;
+    }
+
+    // Loader heap provided at construction time. May be NULL (in which case m_pModule must *not* be NULL).
+    LoaderHeap             *m_pHeap;
+
+    // Fields related to the runtime (volatile or warm) part of the hash.
```
Comment (jkotas) [inline src/coreclr/vm/dacenumerablehash.h:212]:
```suggestion
```

File: src/coreclr/vm/dacenumerablehash.h (hunk 2 above)
Comment (jkotas) [reply src/coreclr/vm/dacenumerablehash.h:215]: (There are a few more mentions of warm to cleanup in the .inl file.)

**Review by jkotas — APPROVED** [review]
Thank you!

**Pull request merged**